### ISSUE TITLE
MultiLabel Metric Support

### ISSUE THREAD

**Issue by SumNeuron** [tensorflow/tensorflow#33636]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.15+ 
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

# Issue
In my comments on [issue](https://github.com/tensorflow/tensorflow/issues/28074) #28074 I outline that although it was suggested multi-label metric support exists, it does not. The reasons stands as

1. The utility function `update_confusion_matrix_variables` is called without a user-exposed option resulting in the default argument `multi_label=False` to be used.

2. `tf.keras.metrics.Metric` generally take in `y_pred` and `y_true`. Multi-label classification might use a loss like `tf.nn.sigmoid_cross_entropy_with_logits(labels=labels, logits=logits)`, which applies a sigmoid during the loss calculation. This suggests that the user should not have the output (logits) of their model undergo a sigmoid transformation, as the loss would then be calculated using `sigmoid(sigmoid(...))`.  While a putative solution to is to make a multi-tailed model (one with and one without the sigmoid transformation), a collective guide on these more nuanced use cases does not exist.

pseudo-code of the solution

```python

inputs = ...
y = tf.keras.layer.SomeLayer(...)(inputs)
y = tf.keras.layer.SomeLayer(...)(y)
output = tf.sigmoid(y)


model = tf.keras.Model(inputs=inputs, outputs={
    'logits': output,
    'raw_out_for_loss': y
})

model.compile(
    ...
    metrics={
        'logits': [MultiLabelMetric(...), ...], 
        'raw_out_for_loss': [],
    }
```

This issue does not arise in the [image segmentation guide](https://www.tensorflow.org/tutorials/images/segmentation#define_the_model) as the labels are no independent, so `sparse_categorical_crossentropy` loss is used. As reflected in the corresponding [`tf.keras.losses.SparseCategoricalCrossentropy` documentation](https://www.tensorflow.org/api_docs/python/tf/keras/losses/SparseCategoricalCrossentropy), while multiple labels exists, the input is expect to confirm to a single label.

Mini-rant aside, even if a user does use a multi-tail model to address this, it does not get around that the current `tf.keras.metrics` do not expose the `multi_label` option, despite it existing in the underlying utility code.

# Feature

At bare minimum expose `multi_label` option to relevant metrics (e.g. `Precision`). 

It would also be nice if a transformation could be applied to the automatically feed in `y_pred` (e.g. sigmoid transformation), so that a multi-tailed model is not required. This might be done with an optional `transformation` argument which will accept a label.

## other option
move multilabel metrics to separate classes, e.g. 

```python
class MultiLabelMacroSpecificity(tf.keras.metrics.Metric):
    
    def __init__(self, name='multi_label_macro_specificity', threshold=0.5, **kwargs):        
        super(MultiLabelMacroSpecificity, self).__init__(name=name, **kwargs)
        self.specificity = self.add_weight(name='mlm_spec', initializer='zeros')        
        self.threshold       = tf.constant(threshold)

        # replace this with tf confusion_matrix utils
        self.true_negatives  = self.add_weight(name='tn', initializer='zeros')
        self.false_positives = self.add_weight(name='fp', initializer='zeros')
    
    def update_state(self, y_true, y_pred):
        
        # Compare predictions and threshold.        
        pred_is_pos  = tf.greater(tf.cast(y_pred, tf.float32), self.threshold)            
        pred_is_neg  = tf.logical_not(tf.cast(pred_is_pos, tf.bool))
        # |-- in case of soft labeling        
        label_is_pos = tf.greater(tf.cast(y_true, tf.float32), self.threshold)                
        label_is_neg = tf.logical_not(tf.cast(label_is_pos, tf.bool))
        
        self.true_negatives.assign_add(tf.reduce_sum(tf.cast(tf.logical_and(pred_is_neg, label_is_neg), tf.float32)))
        self.false_positives.assign_add(
            tf.reduce_sum(tf.cast(tf.logical_and(pred_is_pos, label_is_neg), tf.float32))
        )
        
        tn = self.true_negatives
        fp = self.false_positives
        specificity = tf.div_no_nan(tn, tf.add(tn, fp))
        self.specificity.assign(specificity)
        return specificity
    
    def result(self):
        return self.specificity        

```

**Will this change the current api? How?**
Yes, but barely.  Expose at least `multi_label` on relevant metrics and maybe add / expose a `transformation` argument.

**Who will benefit with this feature?**
If bare minimum is done, those requiring multi-label metrics but do not which to write their own function / subclass `tf.keras.metric.Metric`.

If `transformation` is also added to all metrics, then anyone using tf metrics could make existing metrics work with minimum overhead instead of redesigning their model to meet metric requirements.

**Any Other info.**

**Comment by pavithrasv** [comment]
Adding this to contributions welcome list. Please feel free to add the `multi-label` option to the APIs and send me a PR.

**Comment by SumNeuron** [comment]
@pavithrasv  see above

**Comment by Saqhas** [comment]
@SumNeuron  @pavithrasv Are the functionality for multi-label precison, recall, f1-score, etc are added in tensorflow main code base. 
This is also a issue in tensorflow addon repo. 
Please confirm.
Thank you

**Comment by SumNeuron** [comment]
@Saqhas the code is on a branch in the add-on repo, there is an issue with readme?
https://github.com/tensorflow/addons/pull/976

**Comment by Saqhas** [comment]
@SumNeuron Got it.
So it is still pending to be merged in the main code in the addons repo. But is yet to be merged.
Is there any development still pending in this feature? I want to help
Also why is this feature only going in the addons repo because addons dont have precision, recall and f1score in the first place. Precision and recall are present here. So shouldn't it this feature bee added to both the repos. So that all the metrics present in tensorflow and addons can behave appropriately in both single and multi label scenarios. Like the multilabel class idea you suggested.

**Comment by SumNeuron** [comment]
@Saqhas honestly, no idea. I also just wanted to help but the process is... convoluted. As you can see here (https://github.com/tensorflow/tensorflow/pull/36315#issuecomment-579859684) following the contributing guideline I tried to first add to main repo, but was the later instructed to go to add ons, where it has been stuck.

**Comment by Saqhas** [comment]
Ya i saw that but this is not addition of a new metrics. It is adding feature in the existing metrics. And it is used extensively.
Currently for multilabel there is hardly any metrics that can show the true performance of the model.

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.
